File: kotlin-app/composeApp/src/commonMain/kotlin/app/madar/ui/RealtimeAlertBanner.kt
package app.madar.ui

import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.draw.scale
import androidx.compose.ui.graphics.TransformOrigin
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.sin

// In-app realtime alert — the visual companion to the OS notification (the rebuild
// of the Flutter `NewOrderBanner`, generalized to every alerting event). Driven by
// the core's dedup'd alert (RealtimePlayer.postNotification → AppModel.showRealtimeAlert).
// PERSISTENT (stays until dismissed) and rendered like the iOS notification stack:
// a compact collapsed DECK (top card full, the rest peeking behind) that expands
// to a scrollable list on tap — so it overlays the app, never pushes content down,
// and stays small no matter how many pile in.

/** One in-app alert. `tag` is the core's `"{event_type}:{id}"`, so its prefix
 *  picks the icon (delivery / kitchen / ticket / ready). */
data class RealtimeAlertData(val id: Int, val title: String, val body: String, val tag: String) {
    private val eventType: String get() = tag.substringBefore(':')
    val icon: String get() = when {
        eventType.contains("ready") -> "bell.fill"
        eventType.startsWith("delivery") -> "bicycle"
        eventType.startsWith("kitchen") -> "flame.fill"
        eventType.startsWith("ticket") -> "fork.knife"
        else -> "bell.fill"
    }
}

@Composable
fun RealtimeAlertStack(alerts: List<RealtimeAlertData>, onDismiss: (Int) -> Unit, modifier: Modifier = Modifier) {
    if (alerts.isEmpty()) return
    var expanded by remember { mutableStateOf(false) }
    // Nothing left to fan out → fall back to the deck.
    if (alerts.size <= 1 && expanded) expanded = false

    Box(modifier.widthIn(max = 480.dp).fillMaxWidth().padding(Space.md)) {
        if (expanded) {
            ExpandedAlertList(alerts, onDismiss, onCollapse = { expanded = false })
        } else {
            CollapsedAlertDeck(alerts, onDismiss, onExpand = { if (alerts.size > 1) expanded = true })
        }
    }
}

/** iOS-style collapsed deck: the newest alert sits on top at full size; up to two
 *  behind it peek out — scaled down, nudged down, and dimmed — exactly like a
 *  stacked notification group. Tap to fan the stack out. */
@Composable
private fun CollapsedAlertDeck(alerts: List<RealtimeAlertData>, onDismiss: (Int) -> Unit, onExpand: () -> Unit) {
    val deck = alerts.take(3)
    val multi = alerts.size > 1
    Box(Modifier.fillMaxWidth()) {
        // Draw back-to-front so the newest (index 0) lands on top.
        for (d in deck.indices.reversed()) {
            val a = deck[d]
            val front = d == 0
            Box(
                Modifier.fillMaxWidth().graphicsLayer {
                    val s = 1f - 0.05f * d
                    scaleX = s; scaleY = s
                    translationY = 10.dp.toPx() * d
                    alpha = 1f - 0.22f * d
                    transformOrigin = TransformOrigin(0.5f, 0f) // shrink from the top edge → peeks below
                },
            ) {
                RealtimeAlertCard(
                    a, onDismiss,
                    animated = front, showClose = front,
                    onTap = if (front && multi) onExpand else null,
                )
            }
        }
    }
}

/** Fanned-out, scrollable list (newest on top) + a chevron-up to re-collapse.
 *  Capped generously and scrolls — it never pushes the app down (it's an overlay). */
@Composable
private fun ExpandedAlertList(alerts: List<RealtimeAlertData>, onDismiss: (Int) -> Unit, onCollapse: () -> Unit) {
    val c = madarColors()
    BoxWithConstraints(Modifier.fillMaxWidth()) {
        val maxH = maxHeight * 0.8f
        LazyColumn(
            Modifier.fillMaxWidth().heightIn(max = maxH),
            verticalArrangement = Arrangement.spacedBy(Space.sm),
        ) {
            items(alerts, key = { it.id }) { a ->
                RealtimeAlertCard(a, onDismiss, modifier = Modifier.animateItem())
            }
            item("collapse") {
                Row(Modifier.fillMaxWidth().padding(top = Space.xs), horizontalArrangement = Arrangement.Center) {
                    Box(
                        Modifier.clip(CircleShape).background(c.surfaceAlt)
                            .border(1.dp, c.border, CircleShape)
                            .clickable(interactionSource = remember { MutableInteractionSource() }, indication = null) { onCollapse() }
                            .padding(horizontal = Space.lg, vertical = 6.dp),
                        contentAlignment = Alignment.Center,
                    ) { MadarIcon("chevron.up", tint = c.textSecondary, size = 16.dp) }
                }
            }
        }
    }
}

/** The banner card itself — also used by the screenshot harness for previews.
 *  [animated] runs the looping bounce/wiggle; the peeking deck cards turn it off.
 *  [onTap] (when set) handles a body tap (e.g. expand the deck); the ✕ dismisses. */
@Composable
fun RealtimeAlertCard(
    alert: RealtimeAlertData,
    onDismiss: (Int) -> Unit,
    modifier: Modifier = Modifier,
    animated: Boolean = true,
    showClose: Boolean = true,
    onTap: (() -> Unit)? = null,
) {
    val c = madarColors()
    val haptics = rememberHaptics()
    val shape = RoundedCornerShape(Radii.md)

    // Looping bounce + wiggle (the Flutter LoopingIcon, ported): scale pulses on
    // |sin| and the glyph wiggles on a faster sin. Disabled for peeking deck cards.
    val scale: Float
    val angleDeg: Float
    if (animated) {
        val transition = rememberInfiniteTransition(label = "alert")
        val p by transition.animateFloat(
            initialValue = 0f, targetValue = 1f,
            animationSpec = infiniteRepeatable(tween(900, easing = LinearEasing), RepeatMode.Restart),
            label = "p",
        )
        scale = 1f + 0.18f * abs(sin(p * 2f * PI.toFloat()))
        angleDeg = 0.18f * sin(p * 4f * PI.toFloat()) * (180f / PI.toFloat())
    } else {
        scale = 1f; angleDeg = 0f
    }

    Row(
        modifier.fillMaxWidth()
            .elevation(Elevation.RAISED, shape)
            .clip(shape)
            .background(c.accentBg)
            .border(1.dp, c.accent.copy(alpha = 0.35f), shape)
            .clickable(interactionSource = remember { MutableInteractionSource() }, indication = null) {
                haptics.selection(); onTap?.invoke() ?: onDismiss(alert.id)
            }
            .padding(horizontal = 14.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Space.md),
    ) {
        // Icon over a pulsing glow ring.
        Box(contentAlignment = Alignment.Center) {
            Box(Modifier.size(34.dp).scale(scale).clip(CircleShape).background(c.accent.copy(alpha = Opacity.subtle)))
            MadarIcon(alert.icon, tint = c.accent, size = 20.dp, modifier = Modifier.scale(scale).rotate(angleDeg))
        }
        Column(Modifier.weight(1f)) {
            Text(
                alert.title, color = c.accent, fontFamily = LocalMadarFont.current,
                fontWeight = FontWeight.ExtraBold, fontSize = 13.sp, maxLines = 1, overflow = TextOverflow.Ellipsis,
            )
            if (alert.body.isNotBlank()) {
                Text(
                    alert.body, color = c.textSecondary, fontFamily = LocalMadarFont.current,
                    fontSize = 11.sp, maxLines = 1, overflow = TextOverflow.Ellipsis,
                )
            }
        }
        if (showClose) {
            Box(
                Modifier.size(28.dp).clip(CircleShape)
                    .clickable(interactionSource = remember { MutableInteractionSource() }, indication = null) { onDismiss(alert.id) },
                contentAlignment = Alignment.Center,
            ) { MadarIcon("xmark", tint = c.textMuted, size = 16.dp) }
        }
    }
}

File: swift-app/Sources/MadarUI/RealtimeAlertBanner.swift
// In-app realtime alert — the visual companion to the OS notification (the
// rebuild of the Flutter `NewOrderBanner`, generalized to every alerting event).
// Driven by the core's dedup'd alert (RealtimeAlertPlayer.postNotification →
// AppModel.showRealtimeBanner). PERSISTENT (stays until dismissed) and rendered
// like the iOS notification stack: a compact collapsed DECK (top card full, the
// rest peeking behind) that expands to a scrollable list on tap — so it overlays
// the app, never pushes content down, and stays small no matter how many pile in.
import SwiftUI

/// One in-app alert. `tag` is the core's `"{event_type}:{id}"`, so its prefix
/// picks the icon (delivery / kitchen / ticket / ready).
struct RealtimeAlert: Identifiable, Equatable {
    let id: Int
    let title: String
    let body: String
    let tag: String

    var icon: String {
        let type = tag.split(separator: ":").first.map(String.init) ?? ""
        if type.contains("ready") { return "bell.fill" }
        if type.hasPrefix("delivery") { return "bicycle" }
        if type.hasPrefix("kitchen") { return "flame.fill" }
        if type.hasPrefix("ticket") { return "fork.knife" }
        return "bell.fill"
    }
}

extension View {
    /// Mount the top-anchored realtime alert stack over the app root.
    func realtimeAlertHost(_ app: AppModel) -> some View { modifier(RealtimeAlertHostModifier(app: app)) }
}

private struct StackHeightKey: PreferenceKey {
    static var defaultValue: CGFloat = 0
    static func reduce(value: inout CGFloat, nextValue: () -> CGFloat) { value = max(value, nextValue()) }
}

private struct RealtimeAlertHostModifier: ViewModifier {
    @ObservedObject var app: AppModel
    @State private var expanded = false
    @State private var contentHeight: CGFloat = 0

    func body(content: Content) -> some View {
        content.overlay(alignment: .top) {
            if !app.realtimeAlerts.isEmpty {
                Group {
                    if expanded { expandedList } else { collapsedDeck }
                }
                .frame(maxWidth: 480)
                .padding(.horizontal, Space.md)
                .padding(.top, Space.sm)
                .transition(.move(edge: .top).combined(with: .opacity))
                .zIndex(40)
            }
        }
        .animation(.spring(response: 0.34, dampingFraction: 0.9), value: app.realtimeAlerts)
        .animation(.spring(response: 0.34, dampingFraction: 0.9), value: expanded)
        .onChange(of: app.realtimeAlerts.count) { count in if count <= 1 { expanded = false } }
    }

    /// iOS-style collapsed deck: newest on top at full size; up to two behind peek
    /// out — scaled down, nudged down, dimmed. Tap to fan it out.
    private var collapsedDeck: some View {
        let deck = Array(app.realtimeAlerts.prefix(3).enumerated())
        return ZStack(alignment: .top) {
            // Render deepest first so the newest (depth 0) lands on top.
            ForEach(deck.reversed(), id: \.element.id) { (depth, alert) in
                RealtimeAlertCard(
                    alert: alert,
                    animated: depth == 0,
                    showClose: depth == 0,
                    onTap: app.realtimeAlerts.count > 1 ? { expanded = true } : nil,
                    onDismiss: { app.dismissRealtimeAlert(alert.id) }
                )
                .scaleEffect(1 - 0.05 * CGFloat(depth), anchor: .top)
                .offset(y: 10 * CGFloat(depth))
                .opacity(1 - 0.22 * Double(depth))
                .zIndex(Double(3 - depth))
            }
        }
    }

    /// Fanned-out, scrollable list (newest on top) + a chevron-up to re-collapse.
    /// Sized to content up to a cap, then scrolls — it never pushes the app down.
    private var expandedList: some View {
        ScrollView {
            VStack(spacing: Space.sm) {
                ForEach(app.realtimeAlerts) { alert in
                    RealtimeAlertCard(
                        alert: alert, animated: true, showClose: true, onTap: nil,
                        onDismiss: { app.dismissRealtimeAlert(alert.id) }
                    )
                    .transition(.move(edge: .top).combined(with: .opacity))
                }
                Button { expanded = false } label: {
                    MadarIcon("chevron.up", size: 16).foregroundStyle(Color.secondary)
                        .padding(.horizontal, Space.lg).padding(.vertical, 6)
                        .background(Capsule().fill(.thinMaterial))
                }.buttonStyle(.plain)
            }
            .background(GeometryReader { g in
                Color.clear.preference(key: StackHeightKey.self, value: g.size.height)
            })
        }
        .frame(height: min(contentHeight, 440))
        .onPreferenceChange(StackHeightKey.self) { contentHeight = $0 }
    }
}

struct RealtimeAlertCard: View {
    let alert: RealtimeAlert
    var animated: Bool = true
    var showClose: Bool = true
    var onTap: (() -> Void)? = nil
    let onDismiss: () -> Void
    @Environment(\.theme) private var theme

    var body: some View {
        HStack(spacing: Space.md) {
            iconView.frame(width: 34, height: 34)
            VStack(alignment: .leading, spacing: 1) {
                Text(alert.title).font(.ui(13, .heavy)).foregroundStyle(theme.colors.accent).lineLimit(1)
                if !alert.body.isEmpty {
                    Text(alert.body).font(.ui(11)).foregroundStyle(theme.colors.textSecondary).lineLimit(1)
                }
            }
            Spacer(minLength: Space.sm)
            if showClose {
                Button(action: onDismiss) {
                    MadarIcon("xmark", size: 16).foregroundStyle(theme.colors.textMuted)
                        .frame(width: 28, height: 28)
                }.buttonStyle(.plain)
            }
        }
        .padding(.horizontal, 14).padding(.vertical, 12)
        .background(theme.colors.accentBg)
        .clipShape(RoundedRectangle(cornerRadius: Radii.md, style: .continuous))
        .overlay(
            RoundedRectangle(cornerRadius: Radii.md, style: .continuous)
                .strokeBorder(theme.colors.accent.opacity(0.35), lineWidth: 1)
        )
        .elevation(.raised)
        .contentShape(Rectangle())
        .onTapGesture { Haptics.selection(); (onTap ?? onDismiss)() }
    }

    // Looping bounce + wiggle (the Flutter LoopingIcon, ported); static for the
    // peeking deck cards.
    @ViewBuilder private var iconView: some View {
        if animated {
            TimelineView(.animation) { tl in
                let p = tl.date.timeIntervalSinceReferenceDate.truncatingRemainder(dividingBy: 0.9) / 0.9
                iconStack(scale: 1 + 0.18 * abs(sin(p * 2 * .pi)), angle: 0.18 * sin(p * 4 * .pi))
            }
        } else {
            iconStack(scale: 1, angle: 0)
        }
    }

    private func iconStack(scale: CGFloat, angle: Double) -> some View {
        ZStack {
            Circle().fill(theme.colors.accent.opacity(Opacity.subtle))
                .frame(width: 34, height: 34).scaleEffect(scale)
            MadarIcon(alert.icon, size: 20).foregroundStyle(theme.colors.accent)
                .scaleEffect(scale).rotationEffect(.radians(angle))
        }
    }
}
